synapse/handlers/room_summary.py

```diff
@@ -583,7 +590,9 @@ async def get_federation_hierarchy(
 
         # Iterate through each child and potentially add it, but not its children,
         # to the response.
-        for child_room in root_room_entry.children_state_events:
+        for child_room in itertools.islice(
+            root_room_entry.children_state_events, MAX_ROOMS_PER_SPACE
+        ):
```

Comment:
This changes the `/hierarchy` federation API in a not great way, it would now return:

* The requested room with all of the children state (so references to potentially unlimited).
* The summary of the first 50 rooms.

This would cause the requesting server to query an unknown children rooms individually over federation.

The spec doesn't mention anything about this, but the MSC does:

> Similarly, if a server-set limit on the size of the response is reached, additional rooms are not added to the response and can be queried individually.